superpoint/models/super_point.py

```diff
@@ -0,0 +1,203 @@
+
+    def _detector_loss(self, keypoint_map_labels, logits, **config):
+        labels = tf.cast(tf.expand_dims(keypoint_map_labels, axis=3), tf.float32)
+        labels = tf.space_to_depth(labels, config['grid_size'], data_format='NHWC')
+        shape = tf.shape(labels)
+        shape = tf.concat([shape[:3], [1]], axis=0)
+        labels = tf.concat([2*labels, tf.ones(shape)], 3)  # hacky
+        labels = tf.argmax(labels, axis=3)
+
+        # Apply the cross entropy
+        with tf.device('/cpu:0'):
+            cross_entropy = tf.losses.sparse_softmax_cross_entropy(labels=labels,
+                                                                   logits=logits)
+        return tf.reduce_mean(cross_entropy)
+
+    def _descriptors_loss(self, descriptors, warped_descriptors, H, **config):
+        # Compute the position of the center pixel of every cell in the image
+        (batch_size, Hc, Wc) = descriptors.shape.as_list()[:3]
+        full_shape = [Hc * config['grid_size'], Wc * config['grid_size']]
+        row_coord_cells = np.array(range(Hc)) * config['grid_size']\
+            + config['grid_size'] // 2
+        col_coord_cells = np.array(range(Wc)) * config['grid_size']\
+            + config['grid_size'] // 2
+        row_coord_cells = np.repeat(row_coord_cells, Wc).reshape((Hc * Wc, 1))
+        col_coord_cells = np.tile(col_coord_cells, Hc).reshape((Hc * Wc, 1))
+        coord_cells = tf.Variable(np.concatenate([row_coord_cells, col_coord_cells],
+                                                 axis=1),
+                                  dtype=tf.int32)  # shape = (Hc x Wc, 2)
+        map_cells = tf.scatter_nd(coord_cells,
+                                  tf.ones([Hc * Wc], dtype=tf.float32),
+                                  full_shape)  # shape = (H, W)
+        map_cells = tf.tile(tf.expand_dims(map_cells, 0),
+                            [batch_size, 1, 1])  # shape = (n_batches, H, W)
+
+        # Compute the position of the warped center pixels
+        warped_coord_cells = tf.map_fn(warp_keypoints_to_list,
+                                       (map_cells, H),
+                                       dtype=tf.float32)
+        # shape = (n_batches x Hc x Wc, 2)
+
+        # Reshape the tensors
+        coord_cells = tf.tile(coord_cells,
+                              [batch_size, 1])  # shape = (n_batches x Hc x Wc, 2)
+        coord_cells = tf.reshape(coord_cells, [batch_size, Hc, Wc, 2])
+        warped_coord_cells = tf.reshape(warped_coord_cells, [batch_size, Hc, Wc, 2])
+        coord_cells = tf.cast(tf_repeat(coord_cells, [1, Hc, Wc, 1]), tf.float32)
+        warped_coord_cells = tf.cast(tf.tile(warped_coord_cells, [1, Hc, Wc, 1]),
+                                     tf.float32)
+        # shape = (n_batches, Hc x Hc, Wc x Wc, 2)
+
+        # Compute the pairwise distances and filter the distances less than the threshold
+        cell_distances = tf.norm(coord_cells - warped_coord_cells, axis=-1)
```

Comment:
My guess is that `coord_cells` should be rescaled by 0.5 or the homography should be tweaked.